### ISSUE TITLE
encrypt/decrypt functions became 20x times slower after 24.5

### ISSUE THREAD

**Issue by UnamedRus** [ClickHouse/ClickHouse#65116]
**Company or project name**
Put your company name or project description here

**Describe the situation**
encrypt/decrypt functions

```
24.3.1

SELECT DISTINCT toString(number) = decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), 'key-key-key-key+'), 'key-key-key-key+')
FROM numbers_mt(100000000)

Query id: aaf5862b-4aa0-4508-bdb5-358befadaba2

┌─equals(toString(number), decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), 'key-key-key-key+'), 'key-key-key-key+'))─┐
│                                                                                                                                  1 │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 1.234 sec. Processed 99.29 million rows, 794.33 MB (80.47 million rows/s., 643.77 MB/s.)
Peak memory usage: 83.98 MiB.

SELECT DISTINCT toString(number) = decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), 'key-key-key-key+'), 'key-key-key-key+')
FROM numbers_mt(10000000)
SETTINGS max_threads = 1

┌─equals(toString(number), decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), 'key-key-key-key+'), 'key-key-key-key+'))─┐
│                                                                                                                                  1 │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.829 sec. Processed 9.81 million rows, 78.49 MB (11.84 million rows/s., 94.70 MB/s.)
Peak memory usage: 6.01 MiB.

24.6.1.3534

SELECT DISTINCT toString(number) = decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), 'key-key-key-key+'), 'key-key-key-key+')
FROM numbers_mt(100000000)

   ┌─equals(toString(number), decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), [HIDDEN id: 1]), [HIDDEN id: 1]))─┐
1. │                                                                                                                          1 │
   └────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 23.295 sec. Processed 100.00 million rows, 800.00 MB (4.29 million rows/s., 34.34 MB/s.)
Peak memory usage: 82.05 MiB.



SELECT DISTINCT toString(number) = decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), 'key-key-key-key+'), 'key-key-key-key+')
FROM numbers_mt(10000000)
SETTINGS max_threads = 1

   ┌─equals(toString(number), decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), [HIDDEN id: 1]), [HIDDEN id: 1]))─┐
1. │                                                                                                                          1 │
   └────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 2.824 sec. Processed 9.68 million rows, 77.44 MB (3.43 million rows/s., 27.42 MB/s.)
Peak memory usage: 6.01 MiB.
```

**Expected performance**
What are your performance expectation, why do you think they are realistic? Has it been working faster in older ClickHouse releases? Is it working faster in some specific other system?

**Additional context**

**Comment by qoega** [comment]
We need to add this case to performance tests. They are all green in PR:
https://s3.amazonaws.com/clickhouse-test-reports/59870/9d2301f0f72346d7206debf3e99db8c4217edb91/ci_running.html

**Comment by Enmk** [comment]
Initially, performance tests were added with encrypt/decrypt functions in https://github.com/ClickHouse/ClickHouse/pull/11844 (see `tests/performance/encrypt_decrypt.xml`, and `tests/performance/encrypt_decrypt_slow.xml`)

But later were removed (without clear reason) in https://github.com/ClickHouse/ClickHouse/pull/57494 and https://github.com/clickHouse/clickhouse/commit/7340839d6d016ea70d8cadebe1d871204881e75f (I wasn't able to find a corresponding PR).

Leaving out only `tests/performance/encrypt_decrypt_empty_string.xml`, turns out that wasn't enough to detect performance degradation.

**Comment by rschu1ze** [comment]
Sorry, haven't had time to check yet.

In https://github.com/ClickHouse/ClickHouse/issues/65116#issue-2346826157, the regression is observed with `max_threads = 1`, so it is hard to believe for me to believe this is caused by OpenSSL synchronizing internally on some lock as suggested by https://github.com/openssl/openssl/issues/17064 (upon a quick look).

**Comment by wtarreau** [comment]
Please note that in our tests on haproxy, 3.0 dies under locks, and from 3.1 and upper, they replaced many locks with some atomic ops to offer a slightly "better" scalability. The problem is that there are now so many atomic ops all around the code that everything is constantly much slower than in 1.1.1, even with a single thread. It also has a significant impact on threads running on CPUs with split caches such as AMD EPYC because these atomic ops are causing cache line migrations between these core clusters and this is not particularly fast...

**Comment by UnamedRus** [comment]
@rschu1ze Any update?

What are upstream plans with that? 
It is what it is, and rely on OpenSSL future plans or looking in potential alternatives?

**Comment by qoega** [comment]
@thevar1able did you meant to close this issue? Or it is just a stupid github automation?

**Comment by mkmkme** [comment]
> [@thevar1able](https://github.com/thevar1able) did you meant to close this issue? Or it is just a stupid github automation?

It seems to have been closed by [this](https://github.com/ClickHouse/ClickHouse/pull/99105/changes/88ec546648d5c8174219381178d09412d2a798a3) commit which is a part of #99105

**Comment by rschu1ze** [comment]
@thevar1able Could I ask you to verify if the originally reported regressions are gone? If yes, please re-close this issue again 😄 Thanks.

**Comment by thevar1able** [comment]
Alexey merged https://github.com/ClickHouse/ClickHouse/pull/99105, so this issue can be closed.

Let me recheck if regression is gone just to be sure, and I'll close it.

**Comment by thevar1able** [comment]
Before #99105
```
clickhouse local
ClickHouse local version 26.5.2.39.

:) SELECT DISTINCT toString(number) = decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), 'key-key-key-key+'), 'key-key-key-key+')
FROM numbers_mt(100000000)

SELECT DISTINCT toString(number) = decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), 'key-key-key-key+'), 'key-key-key-key+')
FROM numbers_mt(100000000)

Query id: cd600cd2-3eb7-4e9d-9cad-8e990330cd06

   ┌─equals(toStr⋯EN id: 1]))─┐
1. │                        1 │
   └──────────────────────────┘

1 row in set. Elapsed: 18.855 sec. Processed 100.00 million rows, 800.00 MB (5.30 million rows/s., 42.43 MB/s.)
Peak memory usage: 131.66 MiB.
```

After:
```
./clickhouse
ClickHouse local version 26.6.1.703 (official build).

:) SELECT DISTINCT toString(number) = decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), 'key-key-key-key+'), 'key-key-key-key+')
FROM numbers_mt(100000000)

SELECT DISTINCT toString(number) = decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), 'key-key-key-key+'), 'key-key-key-key+')
FROM numbers_mt(100000000)

Query id: fab3c832-640a-4c64-a306-01301877dea6

   ┌─equals(toStr⋯EN id: 1]))─┐
1. │                        1 │
   └──────────────────────────┘

1 row in set. Elapsed: 10.763 sec. Processed 100.00 million rows, 800.00 MB (9.29 million rows/s., 74.33 MB/s.)
Peak memory usage: 125.60 MiB.
```

So ~1.9x speedup as observed by performance tests. The rest 10x is still missing 🙂 

@rschu1ze I'll leave it up to you to decide if this should be closed.

**Comment by clickgapai** [comment]
_Found via ClickGap automated review. Please close or comment if this is incorrect or needs adjustment._

### TL;DR
**encrypt/decrypt functions became 20x times slower after 24.4**

**ClickGap verified:** reproduced it with the script below on `26.5`; tested 6 version(s) — 5 affected, 1 clean (see Affects); bisected to a candidate range — a single introducing PR could not be pinned (see Bisect).

### Reproduction
```
-- Reporter's roundtrip (fast pre-24.4, ~20x slower on 24.6+ and all 26.x):
SELECT DISTINCT toString(number) = decrypt('aes-128-cbc', encrypt('aes-128-cbc', toString(number), 'key-key-key-key+'), 'key-key-key-key+')
FROM numbers_mt(100000000);
-- Observed: ~1.2s on 24.3.1; ~23s on 24.6+ and on every supported release 26.1-26.5
-- (OpenSSL 3.x re-fetches the cipher under a global lock once per row). Fixed on master by #107339.
```

### Bisect
- The bisect boundary was rejected — PR #99207 only adds a test for AutoPR dataflow cache reuse and does not modify any encryption/decryption function code or OpenSSL integration logic. — a coincidental boundary, not the cause; no introducing PR pinned.
- The source-build bisect converged to #99207, but it is **not** the introducer (see above) — not attributing it.

### Affects
- ✓ **Fix landed on master** in PR #107339. Branches below still have the bug and need the backport.
- **Reproduces on**: `26.5, 26.4, 26.3, 26.2, 26.1`
- **Does NOT reproduce on**: `master`
- **Backport needed**: 26.5, 26.4, 26.3, 26.2, 26.1

### Suggested next step
The bisect boundary above narrows the search in `comp-regular-function` — treat the candidate range as the starting point; the single introducing PR is unconfirmed.

**Component:** `comp-regular-function`
**Severity:** `P2`

---
_ClickGap · Finding: `h_issue_65116`_




### Triage metadata
- Primary component: `comp-regular-function`
- Secondary components: `comp-external-dependencies`
- Component owners (`comp-regular-function`): @alexey-milovidov @bharatnc
